fdbcli/DebugCommands.actor.cpp

```diff
@@ -238,39 +256,106 @@ bool checkResults(Version version,
+						       printStorageServerMachineInfo(servers[j]).c_str(),
 						       currentI,
 						       referenceI,
 						       version,
 						       toHex(currentKV.key).c_str());
+						corruptedKeys.insert(currentKV.key);
 					}
 					currentI++;
 					referenceI++;
 				} else if (currentKV.key < referenceKV.key) {
 					printf("Inconsistency: UniqueKey, %s(1), %s(0), CurrentIndex %lu, ReferenceIndex %lu, Version %ld, "
 					       "Key %s\n",
-					       servers[j].address().toString().c_str(),
-					       servers[firstValidServer].address().toString().c_str(),
+					       printStorageServerMachineInfo(servers[j]).c_str(),
+					       printStorageServerMachineInfo(servers[firstValidServer]).c_str(),
 					       currentI,
 					       referenceI,
 					       version,
 					       toHex(currentKV.key).c_str());
+					corruptedKeys.insert(currentKV.key);
 					currentI++;
 				} else {
 					printf("Inconsistency: UniqueKey, %s(1), %s(0), CurrentIndex %lu, ReferenceIndex %lu, Version %ld, "
 					       "Key %s\n",
-					       servers[firstValidServer].address().toString().c_str(),
-					       servers[j].address().toString().c_str(),
+					       printStorageServerMachineInfo(servers[firstValidServer]).c_str(),
+					       printStorageServerMachineInfo(servers[j]).c_str(),
 					       currentI,
 					       referenceI,
 					       version,
 					       toHex(referenceKV.key).c_str());
+					corruptedKeys.insert(referenceKV.key);
+					referenceI++;
+				}
+			}
+		}
+	}
+
+	if (allSame) {
+		return allSame;
+	}
+
+	firstValidServer = -1;
+	for (int j = 0; j < replies.size(); j++) {
+		if (firstValidServer == -1) {
+			firstValidServer = j;
+			continue; // always select the first server as reference
+		}
+		// compare reference and current
+		GetKeyValuesReply current = replies[j];
+		GetKeyValuesReply reference = replies[firstValidServer];
+		// Detecting corrupted keys for any mismatching replies between current and reference servers
+		size_t currentI = 0, referenceI = 0;
+		while (currentI < current.data.size() || referenceI < reference.data.size()) {
+			if (hasMore && ((referenceI < reference.data.size() && reference.data[referenceI].key >= claimEndKey) ||
+			                (currentI < current.data.size() && current.data[currentI].key >= claimEndKey))) {
+				// If there will be a next round and the key is out of claimEndKey
+				// We will delay the detection to the next round
+				break;
+			}
+			if (currentI >= current.data.size()) {
+				referenceI++;
+			} else if (referenceI >= reference.data.size()) {
+				currentI++;
+			} else {
+				KeyValueRef currentKV = current.data[currentI];
+				KeyValueRef referenceKV = reference.data[referenceI];
+				if (currentKV.key == referenceKV.key) {
+					if (corruptedKeys.find(currentKV.key) != corruptedKeys.end()) {
+						if (currentKV.value == referenceKV.value) {
+							printf("Inconsistency: CorruptedMatch, %s(1), %s(1), CurrentIndex %lu, ReferenceIndex %lu, "
```

Comment:
this suggests to be a exact match(i.e. not a corruption, right?)

Reply:
This is the exact match on a corrupted key

Reply:
yes, this key is corrupted on another SS, but this SS has the correct value(or the same value as reference SS)